backend/src/sms/middlewares/tests/sms-template.middleware.test.ts

```diff
@@ -0,0 +1,156 @@
+  })
+  campaignId = campaign.id
+})
+
+afterEach(async () => {
+  await SmsTemplate.destroy({ where: {} })
+  await SmsMessage.destroy({ where: {} })
+  await Campaign.update({ s3Object: {} }, { where: { id: campaignId } })
+})
+
+afterAll(async () => {
+  await Campaign.destroy({ where: {} })
+  await User.destroy({ where: {} })
+  await sequelize.close()
+  RedisService.otpClient.quit()
+  RedisService.sessionClient.quit()
+})
+
+describe('uploadCompleteHandler middleware', () => {
+  test('Stores error message in s3Object if file is invalid', async () => {
+    await SmsTemplate.create({
+      campaignId: campaignId,
+      params: { variable1: 'abc' },
+      body: 'test {{variable1}}',
+    })
+
+    const mockExtractParamsFromJwt = jest
+      .spyOn(UploadService, 'extractParamsFromJwt')
+      .mockReturnValue({ s3Key: 'key' })
+
+    // mock invalid recipient file
+    const fileStream = Readable.from('name\nabc')
+
+    const s3ClientMock = (S3Client as jest.Mock).mockImplementation(() => {
+      return {
+        download: (): NodeJS.ReadableStream => fileStream,
+      }
+    })
+
+    mockRequest = {
+      session: { user: { id: 1 } } as any,
+      params: { campaignId: String(campaignId) },
+      body: { transaction_id: '123', filename: 'abc', etag: '123' },
+    }
+
+    await SmsTemplateMiddleware.uploadCompleteHandler(
+      mockRequest as Request,
+      mockResponse as Response,
+      nextFunction
+    )
+
+    const campaign = await Campaign.findOne({ where: { id: campaignId } })
+    expect(campaign?.s3Object).toEqual({
+      temp_filename: 'abc',
+      error:
+        "Error: 'recipient' column is missing from the uploaded recipient file. Please check the cell in your uploaded CSV file to ensure the recipient's contact info is correctly labelled as 'recipient'.",
+    })
+
+    mockExtractParamsFromJwt.mockRestore()
+    s3ClientMock.mockRestore()
+  })
+
+  test('Stores filename in s3Object if file is valid', async () => {
+    await SmsTemplate.create({
+      campaignId: campaignId,
+      params: { variable1: 'abc' },
+      body: 'test {{name}}',
+    })
+
+    const mockExtractParamsFromJwt = jest
+      .spyOn(UploadService, 'extractParamsFromJwt')
+      .mockReturnValue({ s3Key: 'key' })
+
+    // mock valid recipient file
+    const fileStream = Readable.from('recipient,name\n8123,abc')
+
+    const s3ClientMock = (S3Client as jest.Mock).mockImplementation(() => {
+      return {
+        download: (): NodeJS.ReadableStream => fileStream,
+      }
+    })
+
+    mockRequest = {
+      session: { user: { id: 1 } } as any,
+      params: { campaignId: String(campaignId) },
+      body: { transaction_id: '123', filename: 'abc', etag: '123' },
+    }
+
+    await SmsTemplateMiddleware.uploadCompleteHandler(
+      mockRequest as Request,
+      mockResponse as Response,
+      nextFunction
+    )
+
+    await new Promise<void>((resolve) =>
+      setTimeout(() => {
+        resolve()
+      }, 1000)
+    )
```

Comment:
The following asserts on the sms messages table being populated fails, I suspect due to the async `parseAndProcessCsv` not having completed running. However this sleep function does not seem to resolve the issue. Similar issue for the telegram template middleware.